src/test/abi_component/widget.cpp

```diff
@@ -0,0 +1,145 @@
+#include "iwidget.h"
+#include <pal_error.h>
+#include <string_view>
+
+struct widget : iwidget
+{
+    int32_t XLANG_CALL QueryInterface(xlang_guid const& id, void** object) noexcept final
+    {
+        if (id == xlang_unknown_guid)
+        {
+            static_assert(std::is_base_of_v<xlang_unknown, iwidget>, "Can only combine these two cases is this is true.");
+            *object = static_cast<xlang_unknown*>(this);
+        }
+        else if (id == iwidget_guid)
+        {
+            *object = static_cast<iwidget*>(this);
+        }
+        else
+        {
+            *object = nullptr;
+            return xlang_error_no_interface;
+        }
+        AddRef();
+        return 0;
+    }
+
+    uint32_t XLANG_CALL AddRef() noexcept final
+    {
+        return ++m_count;
+    }
+
+    uint32_t XLANG_CALL Release() noexcept final
+    {
+        auto result = --m_count;
+        if (result == 0)
+        {
+            delete this;
+        }
+        return result;
+    }
+
+    xlang_error_info* get_answer(int32_t* answer) noexcept override
+    {
+        if (!answer)
+        {
+            return xlang::originate_error(xlang_error_pointer);
+        }
+
+        *answer = 42;
+        return nullptr; // xlang_error_ok;
+    }
+
+    private:
+        xlang::impl::atomic_ref_count m_count;
+};
+
+struct widget_factory : iwidget_factory
+{
+    int32_t XLANG_CALL QueryInterface(xlang_guid const& id, void** object) noexcept final
+    {
+        if (id == xlang_unknown_guid)
+        {
+            static_assert(std::is_base_of_v<xlang_unknown, iwidget_factory>, "Can only combine these two cases is this is true.");
+            *object = static_cast<xlang_unknown*>(this);
+        }
+        else if (id == iwidget_factory_guid)
+        {
+            *object = static_cast<iwidget_factory*>(this);
+        }
+        else
+        {
+            *object = nullptr;
+            return xlang_error_no_interface;
+        }
+        AddRef();
+        return 0;
+    }
+
+    uint32_t XLANG_CALL AddRef() noexcept final
+    {
+        return ++m_count;
+    }
+
+    uint32_t XLANG_CALL Release() noexcept final
+    {
+        auto result = --m_count;
+        if (result == 0)
+        {
+            delete this;
+        }
+        return result;
+    }
+
+    xlang_error_info* activate_widget(iwidget** instance) noexcept override
+    {
+        if (!instance)
```

Comment:
No need to validate this. AV is far better for invalid input.

Reply:
This is the opposite of what @Scottj1s suggested

Reply:
but I agree with it, so I'll remove these unless @Scottj1s raises a stink